Introduced FlowLogic.getFlowContext which provides the flow version a…

shamsasari opened the pull request:
…nd app name of the other side.

I decided to change the app version to app name, which is the CorDapp jar file name, as it should already contain the app version.

mikehearn commented on docs/source/versioning.rst:
```diff
@@ -31,14 +31,39 @@ for the network.
-   once we start loading CorDapps in separate class loaders.
+In addition to the evolution of the platform, flows that run on top of the platform can also evolve. It may be that the
+flow protocol between an initiating flow and it's intiated flow changes from one CorDapp release to the next in such as
+way to be backwards incompatible with existing flows. For example, if and sequence of sends and receives needs to change
```
if and -> if a

mikehearn commented on docs/source/versioning.rst:
```diff
@@ -31,14 +31,39 @@ for the network.
+
+The above shows an example evolution of a flow which in the first version was expecting to receive an Int, but then
+in subsequent versions was relaxed to receive a String. This flow is still able to communicate with parties which are
+running the older flow (or rather older CorDapps containing the older flow).
```
Can we have some text about the app name here?

mikehearn commented on node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt:
```diff
@@ -7,36 +7,37 @@ import net.corda.core.internal.castIfPossible
+
+data class SessionInit(val initiatorSessionId: Long,
+                       val initiatingFlowClass: String,
+                       val flowVerison: Int,
```
typo: flowVersion

mikehearn commented on node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt:
```diff
@@ -7,36 +7,37 @@ import net.corda.core.internal.castIfPossible
+
+data class SessionConfirm(override val initiatorSessionId: Long,
+                          val initiatedSessionId: Long,
+                          val flowVerison: Int,
```
typo: flowVersion

shamsasari replied:
Gah, last minute refactorings.

shamsasari commented on core/src/main/kotlin/net/corda/core/node/services/NotaryService.kt:
```diff
@@ -21,11 +22,9 @@ abstract class NotaryService : SingletonSerializeAsToken() {
-     * The first parameter is the client [Party] making the request and the second is the platform version
-     * of the client's node. Use this version parameter to provide backwards compatibility if the notary flow protocol
-     * changes.
+     * The lambda parameter is the client [Party] making the request.
```
Oops

shamsasari commented:
I've addressed your comments.

I add some docs to say the CoDapp filename can never change once deployed, to make sure counter-party flows have something stable to reply on. That implies the filename should also be in the MANIFEST file, and the node checks that it matches the filename. WDYT?

shamsasari commented:
Sorry, accidentally closed the PR.

mikehearn commented:
Well, perhaps we don't need a rule that says it can "never" be changed, just that if you change it you should be sure that nobody is checking the appName field in the context object.
